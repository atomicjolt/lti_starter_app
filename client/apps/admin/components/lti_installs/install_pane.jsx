import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import _ from 'lodash';
import AccountInstall from './account_install';
import CourseInstalls from './course_installs';
import Pagination from './pagination';

const PAGE_SIZE = 10;
const COURSE_TYPES = ['basic', 'course_navigation', 'wysiwyg_button'];

function select(state) {
  return {
    allCourses: state.courses,
  };
}

export class InstallPane extends React.Component {
  static propTypes = {
    courses             : PropTypes.arrayOf(PropTypes.shape({})).isRequired,
    loadExternalTools   : PropTypes.func,
    applicationInstance : PropTypes.shape({}),
    canvasRequest       : PropTypes.func,
    loadingCourses      : PropTypes.shape({}),
    account             : PropTypes.shape({
      installCount : PropTypes.number
    }),
    onlyShowInstalled: PropTypes.bool.isRequired,
  };

  constructor() {
    super();
    this.state = {
      currentPage: 0,
      searchPrefix: '',
    };
  }

  componentDidMount() {
    if (!_.isEmpty(this.props.courses)) {
      this.loadExternalTools();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if ((prevState.currentPage !== this.state.currentPage) ||
       (prevProps.courses.length !== this.props.courses.length) ||
       (prevState.searchPrefix !== this.state.searchPrefix)
    ) {
      this.loadExternalTools();
    }
  }

  isSearching() {
    return !_.isEmpty(this.state.searchPrefix);
  }

  filteredCourses(courses) {
    const {
      applicationInstance,
      onlyShowInstalled,
    } = this.props;

    if (onlyShowInstalled) {
      return _.filter(courses, course => (
        _.find(
          course.external_tools,
          tool => tool.consumer_key === applicationInstance.lti_key
        )
      ));
    }
    return courses;
  }

  searchedCourses() {
    const { allCourses } = this.props;
    return _.filter(this.filteredCourses(allCourses), course => (
      _.includes(
        _.lowerCase(course.name),
        _.lowerCase(this.state.searchPrefix)
      )
    ));
  }

  courses() {
    const { courses } = this.props;
    return _.filter(this.filteredCourses(courses), course => (
      _.includes(
        _.lowerCase(course.name),
        _.lowerCase(this.state.searchPrefix)
      )
    ));
  }

  pageCourses(searchedCourses) {
    return _.slice(
      searchedCourses,
      this.state.currentPage * PAGE_SIZE,
     (this.state.currentPage * PAGE_SIZE) + PAGE_SIZE
   );
  }

  loadExternalTools() {
    _.each(this.pageCourses(this.searchedCourses()), (course) => {
      if (course.external_tools === undefined) {
        this.props.loadExternalTools(course.id);
      }
    });
  }

  updateSearchPrefix = _.debounce((searchPrefix) => {
    this.setState({ searchPrefix, currentPage: 0 });
  }, 150)

  render() {
    const searchedCourses =  this.isSearching() ? this.searchedCourses() : this.courses();
    const pageCount = _.ceil(searchedCourses.length / PAGE_SIZE);
    let accountInstall = null;
    let courseInstalls = null;
    const {
      applicationInstance,
    } = this.props;

    if (applicationInstance) {
      accountInstall = (
        <AccountInstall
          account={this.props.account}
          accountInstalls={this.props.account ? this.props.account.installCount : null}
          applicationInstance={applicationInstance}
          canvasRequest={this.props.canvasRequest}
        />
      );
      courseInstalls = (
        <div>
          <div className="c-search c-search--small">
            <input
              type="text"
              placeholder="Search..."
              onChange={e => this.updateSearchPrefix(e.target.value)}
            />
            <i className="i-search" />
          </div>
          {
            !_.isEmpty(this.props.loadingCourses) ?
              <div className="c-modal--error loading">
                <div className="c-loading-icon" />
              </div> : null
          }
          <CourseInstalls
            applicationInstance={applicationInstance}
            courses={this.pageCourses(searchedCourses)}
            loadingCourses={this.props.loadingCourses}
            canvasRequest={this.props.canvasRequest}
          />
          <Pagination
            setPage={change => this.setState({ currentPage: change.selected })}
            pageCount={pageCount}
            courses={this.props.courses}
            pageSize={PAGE_SIZE}
            loadingCourses={this.props.loadingCourses}
            currentPage={this.state.currentPage}
          />
        </div>
      );
    }
    return (
      <div className="o-right">
        {accountInstall}
        {courseInstalls}
      </div>
    );
  }
}

export default connect(select, {})(InstallPane);
