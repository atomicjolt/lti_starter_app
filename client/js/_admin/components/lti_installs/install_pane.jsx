import React          from 'react';
import _              from 'lodash';
import AccountInstall from './account_install';
import CourseInstalls from './course_installs';
import Pagination     from './pagination';

const PAGE_SIZE = 10;
const COURSE_TYPES = ['basic', 'course_navigation'];

export default class InstallPane extends React.Component {
  static propTypes = {
    courses             : React.PropTypes.arrayOf(React.PropTypes.shape({})).isRequired,
    loadExternalTools   : React.PropTypes.func,
    applicationInstance : React.PropTypes.shape({
      lti_type: React.PropTypes.string.isRequired,
    }),
    canvasRequest       : React.PropTypes.func,
    loadingCourses      : React.PropTypes.shape({}),
    account             : React.PropTypes.shape({
      installCount : React.PropTypes.number
    }),
  };

  constructor() {
    super();
    this.state = {
      currentPage  : 0,
      searchPrefix : '',
    };
  }

  componentDidMount() {
    if (!_.isEmpty(this.props.courses)) {
      this.loadExternalTools();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if ((prevState.currentPage !== this.state.currentPage) ||
       (prevProps.courses.length !== this.props.courses.length) ||
       (prevState.searchPrefix !== this.state.searchPrefix)
    ) {
      this.loadExternalTools();
    }
  }

  searchedCourses() {
    return _.filter(this.props.courses, course => (
      _.includes(
        _.lowerCase(course.name),
        _.lowerCase(this.state.searchPrefix)
      )
    ));
  }

  pageCourses(searchedCourses) {
    return _.slice(
      searchedCourses,
      this.state.currentPage * PAGE_SIZE,
     (this.state.currentPage * PAGE_SIZE) + PAGE_SIZE
   );
  }

  loadExternalTools() {
    _.each(this.pageCourses(this.searchedCourses()), (course) => {
      if (course.external_tools === undefined) {
        this.props.loadExternalTools(course.id);
      }
    });
  }

  updateSearchPrefix = _.debounce((searchPrefix) => {
    this.setState({ searchPrefix, currentPage: 0 });
  }, 150)

  render() {
    const searchedCourses = this.searchedCourses();
    const pageCount = _.ceil(searchedCourses.length / PAGE_SIZE);
    let accountInstall = null;
    let courseInstalls = <p className="c-alert c-alert--info">Course install not available for this tool</p>;
    const {
      applicationInstance,
    } = this.props;

    if (applicationInstance) {
      accountInstall = (
        <AccountInstall
          account={this.props.account}
          accountInstalls={this.props.account ? this.props.account.installCount : null}
          applicationInstance={applicationInstance}
          canvasRequest={this.props.canvasRequest}
        />
      );
      if (_.includes(COURSE_TYPES, applicationInstance.lti_type)) {
        courseInstalls = (
          <span>
            <div className="c-search c-search--small">
              <input
                type="text"
                placeholder="Search..."
                onChange={e => this.updateSearchPrefix(e.target.value)}
              />
              <i className="i-search" />
            </div>
            {
              !_.isEmpty(this.props.loadingCourses) ?
                <div className="c-modal--error loading">
                  <div className="c-loading-icon" />
                </div> : null
            }
            <CourseInstalls
              applicationInstance={applicationInstance}
              courses={this.pageCourses(searchedCourses)}
              loadingCourses={this.props.loadingCourses}
              canvasRequest={this.props.canvasRequest}
            />
            <Pagination
              setPage={change => this.setState({ currentPage: change.selected })}
              pageCount={pageCount}
              courses={this.props.courses}
              pageSize={PAGE_SIZE}
              loadingCourses={this.props.loadingCourses}
              currentPage={this.state.currentPage}
            />
          </span>
        );
      }
    }
    return (
      <div className="o-right">
        {accountInstall}
        {courseInstalls}
      </div>
    );
  }
}
