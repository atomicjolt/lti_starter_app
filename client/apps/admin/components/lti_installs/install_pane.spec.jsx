import React from 'react';
import TestUtils from 'react-dom/test-utils';
import Stub from '../../../../specs_support/stub';
import { InstallPane } from './install_pane';

describe('install pane', () => {
  let result;

  const props = {
    canvasRequest: () => {},
    loadingCourses: {},
    applicationInstance: {},
    courses: [],
    account: {
      installCount: 0,
    },
    loadExternalTools: () => {},
    onlyShowInstalled: false,
  };

  describe('basic instances', () => {
    beforeEach(() => {
      result = TestUtils.renderIntoDocument(
        <Stub>
          <InstallPane {...props} />
        </Stub>
      );
    });

    it('renders the install pane with account installs for basic', () => {
      const button = TestUtils.findRenderedDOMComponentWithTag(result, 'button');
      expect(button.textContent).toContain('Install Into Account');
    });

    it('renders the install pane with course installs for basic', () => {
      const input = TestUtils.findRenderedDOMComponentWithTag(result, 'input');
      expect(input.placeholder).toContain('Search...');
    });
  });

});
