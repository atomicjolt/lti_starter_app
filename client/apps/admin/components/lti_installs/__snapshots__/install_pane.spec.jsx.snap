// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install pane matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <InstallPane
    account={
      Object {
        "installCount": 0,
      }
    }
    applicationInstance={Object {}}
    canvasRequest={[Function]}
    courses={Object {}}
    loadExternalTools={[Function]}
    loadingCourses={Object {}}
    onlyShowInstalled={false}
    onlyShowInstalledChanged={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <AccountInstall
          account={
            Object {
              "installCount": 0,
            }
          }
          accountInstalls={0}
          applicationInstance={Object {}}
          canvasRequest={[Function]}
        />,
        <div>
          <div
            className="c-search c-search--small"
          >
            <input
              onChange={[Function]}
              placeholder="Search..."
              type="text"
            />
            <i
              className="i-search"
            />
          </div>
          <CourseInstalls
            applicationInstance={Object {}}
            canvasRequest={[Function]}
            courses={Array []}
            loadingCourses={Object {}}
            onlyShowInstalledChanged={[Function]}
          />
          <Pagination
            courses={Object {}}
            currentPage={0}
            loadingCourses={Object {}}
            pageCount={0}
            pageSize={10}
            setPage={[Function]}
          />
        </div>,
      ],
      "className": "o-right",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "account": Object {
            "installCount": 0,
          },
          "accountInstalls": 0,
          "applicationInstance": Object {},
          "canvasRequest": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="c-search c-search--small"
            >
              <input
                onChange={[Function]}
                placeholder="Search..."
                type="text"
              />
              <i
                className="i-search"
              />
            </div>,
            null,
            <CourseInstalls
              applicationInstance={Object {}}
              canvasRequest={[Function]}
              courses={Array []}
              loadingCourses={Object {}}
              onlyShowInstalledChanged={[Function]}
            />,
            <Pagination
              courses={Object {}}
              currentPage={0}
              loadingCourses={Object {}}
              pageCount={0}
              pageSize={10}
              setPage={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  onChange={[Function]}
                  placeholder="Search..."
                  type="text"
                />,
                <i
                  className="i-search"
                />,
              ],
              "className": "c-search c-search--small",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Search...",
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "i-search",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
            ],
            "type": "div",
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "applicationInstance": Object {},
              "canvasRequest": [Function],
              "courses": Array [],
              "loadingCourses": Object {},
              "onlyShowInstalledChanged": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "courses": Object {},
              "currentPage": 0,
              "loadingCourses": Object {},
              "pageCount": 0,
              "pageSize": 10,
              "setPage": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <AccountInstall
            account={
              Object {
                "installCount": 0,
              }
            }
            accountInstalls={0}
            applicationInstance={Object {}}
            canvasRequest={[Function]}
          />,
          <div>
            <div
              className="c-search c-search--small"
            >
              <input
                onChange={[Function]}
                placeholder="Search..."
                type="text"
              />
              <i
                className="i-search"
              />
            </div>
            <CourseInstalls
              applicationInstance={Object {}}
              canvasRequest={[Function]}
              courses={Array []}
              loadingCourses={Object {}}
              onlyShowInstalledChanged={[Function]}
            />
            <Pagination
              courses={Object {}}
              currentPage={0}
              loadingCourses={Object {}}
              pageCount={0}
              pageSize={10}
              setPage={[Function]}
            />
          </div>,
        ],
        "className": "o-right",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "account": Object {
              "installCount": 0,
            },
            "accountInstalls": 0,
            "applicationInstance": Object {},
            "canvasRequest": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="c-search c-search--small"
              >
                <input
                  onChange={[Function]}
                  placeholder="Search..."
                  type="text"
                />
                <i
                  className="i-search"
                />
              </div>,
              null,
              <CourseInstalls
                applicationInstance={Object {}}
                canvasRequest={[Function]}
                courses={Array []}
                loadingCourses={Object {}}
                onlyShowInstalledChanged={[Function]}
              />,
              <Pagination
                courses={Object {}}
                currentPage={0}
                loadingCourses={Object {}}
                pageCount={0}
                pageSize={10}
                setPage={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    onChange={[Function]}
                    placeholder="Search..."
                    type="text"
                  />,
                  <i
                    className="i-search"
                  />,
                ],
                "className": "c-search c-search--small",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Search...",
                    "type": "text",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "i-search",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
              ],
              "type": "div",
            },
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "applicationInstance": Object {},
                "canvasRequest": [Function],
                "courses": Array [],
                "loadingCourses": Object {},
                "onlyShowInstalledChanged": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "courses": Object {},
                "currentPage": 0,
                "loadingCourses": Object {},
                "pageCount": 0,
                "pageSize": 10,
                "setPage": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
